src/runtime/virtcontainers/nydusd.go

```diff
@@ -0,0 +1,424 @@
+	keepAlive             = 5 * time.Second
+	expectContinueTimeout = 1 * time.Second
+
+	nydusRafsType          = "rafs"
+	nydusPassthroughfsType = "passthrough_fs"
+
+	sharedPathInGuest = "/containers"
+
+	shimNsPath = "/proc/self/ns/net"
+)
+
+var (
+	nydusdTracingTags = map[string]string{
+		"source":    "runtime",
+		"package":   "virtcontainers",
+		"subsystem": "nydusd",
+	}
+
+	errNydusdDaemonPathEmpty    = errors.New("nydusd daemon path is empty")
+	errNydusdSocketPathEmpty    = errors.New("nydusd socket path is empty")
+	errNydusdAPISocketPathEmpty = errors.New("nydusd api socket path is empty")
+	errNydusdSourcePathEmpty    = errors.New("nydusd source path is empty")
+	errNydusdSourceNotAvailable = errors.New("nydusd source path not available")
+	errNydusdNotSupport         = errors.New("nydusd only support qemu currently")
+)
+
+type nydusd struct {
+	pid             int
+	path            string
+	sockPath        string
+	apiSockPath     string
+	sourcePath      string
+	debug           bool
+	extraArgs       []string
+	startFn         func(cmd *exec.Cmd) error // for mock testing
+	setupShareDirFn func() error              // for mock testing
+}
+
+func startInShimNS(cmd *exec.Cmd) error {
+	// Create nydusd in shim netns as it needs to access host network
+	return doNetNS(shimNsPath, func(_ ns.NetNS) error {
+		return cmd.Start()
+	})
+}
+
+func (nd *nydusd) Start(ctx context.Context, onQuit onQuitFunc) (int, error) {
+	span, _ := katatrace.Trace(ctx, nd.Logger(), "Start", nydusdTracingTags)
+	defer span.End()
+	pid := 0
+
+	if err := nd.valid(); err != nil {
+		return pid, err
+	}
+
+	args, err := nd.args()
+	if err != nil {
+		return pid, err
+	}
+	cmd := exec.Command(nd.path, args...)
+	r, w, err := os.Pipe()
+	if err != nil {
+		return pid, err
+	}
+	cmd.Stdout = w
+	cmd.Stderr = w
+	fields := logrus.Fields{
+		"path": nd.path,
+		"args": strings.Join(args, " "),
+	}
+	nd.Logger().WithFields(fields).Info()
+	if err := nd.startFn(cmd); err != nil {
+		return pid, err
+	}
+	// Monitor nydusd's stdout/stderr and stop sandbox if nydusd quits
+	go func() {
+		scanner := bufio.NewScanner(r)
+		for scanner.Scan() {
+			nd.Logger().Info(scanner.Text())
+		}
+		nd.Logger().Info("nydusd quits")
+		// Wait to release resources of nydusd process
+		_, err = cmd.Process.Wait()
+		if err != nil {
+			nd.Logger().WithError(err).Warn("nydusd quits")
+		}
+		if onQuit != nil {
+			onQuit()
+		}
+	}()
+	if err := nd.setupShareDirFn(); err != nil {
+		return pid, err
+	}
+	nd.pid = cmd.Process.Pid
+	return nd.pid, nil
+}
+
+func (nd *nydusd) args() ([]string, error) {
+	logLevel := "info"
+	if nd.debug {
+		logLevel = "debug"
+	}
+	args := []string{
+		"--log-level", logLevel,
+		"--apisock", nd.apiSockPath,
+		"--sock", nd.sockPath,
+	}
+	if len(nd.extraArgs) > 0 {
+		args = append(args, nd.extraArgs...)
+	}
+	return args, nil
+}
+
+func (nd *nydusd) valid() error {
+	if len(nd.path) == 0 {
+		return errNydusdDaemonPathEmpty
```

Comment:
- What if the value is set but the path doesn't exist (with the correct type)?
- What if the path is not absolute?

Same comments for the other paths below.

Reply:
have updated and add more check. but before I refer to [virtiofsd.go](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/virtiofsd.go#L199), maybe here also need update.